components/tabs/src/storage.rs

```diff
@@ -362,6 +369,25 @@ impl TabsStorage {
     }
 }
 
+// Trim the amount of tabs in a list to fit the specified memory size
+fn trim_tabs_length(tabs: &mut Vec<RemoteTab>, memory_limit: usize) -> Option<Vec<RemoteTab>> {
+    let tabs_mem_size = calc_memory_size(&tabs);
+    // Don't trim if we're under the limit
+    if tabs_mem_size < memory_limit {
+        return Some(tabs.to_vec());
+    }
+    // Calculate the difference needed to hit the memory payload
+    // then drop any over
+    let difference = tabs_mem_size - memory_limit;
+    let elements_to_remove = difference / mem::size_of::<RemoteTab>();
+    tabs.truncate(tabs.len() - elements_to_remove);
+    Some(tabs.to_vec())
+}
+
+fn calc_memory_size<T>(v: &Vec<T>) -> usize {
+    v.capacity() * std::mem::size_of::<T>() + std::mem::size_of::<Vec<T>>()
```

Comment:
> 



> Hmm what does the server reject records on? I imagine it's the size of the encrypted payload json? Are we trying to do best-effort trimming at this layer (i.e we don't need to be accurate?)
> 
> I say that because [`std::mem::size_of`](https://doc.rust-lang.org/std/mem/fn.size_of.html) is the compile time size and:
> 
>     * adding the size of a `Vec` (which in practice is a raw pointer and two `usize`s)
> 
>     * the `std::mem::sizeof::<T>()` is probably not an accurate size, since that is the compile time size. i.e `RemoteTab` is
>       ```rust
>       pub struct RemoteTab {
>         pub title: String,
>         pub url_history: Vec<String>,
>         pub icon: Option<String>,
>         pub last_used: i64, // In ms.
>       }
>       ```
>         
>       since most values there are heap-allocated we could underestimate the size of the actual tab once it's serialized
> 
>     * Using `capacity` instead of `len` may overestimate the size of the vector
> 
> 
> if that is OK, and we aren't aiming for full accuracy (which I think is very reasonable TBH) then it's alright, but we'd want some documentation to convince our future selves what the trimming logic is doing

Yes! As a first pass at this, I definitely wanted to implement something that was (1) easy to maintain, (2) relatively easy to read/understand (though needs more comments I agree), (3) comparatively fast.

re telemetry: I don't think we _know_ exactly how many records are failing to upload, since we don't keep track of that. However, that'll be something I'll also want to implement so we can see (1) implementing telemetry success/failure for individual records (2) then see tabs records failing as a result of "too large" dropping

Reply:
Sorry for the quick comment before, but I think we really just need to reimplement that `tryFitItems` function (ie, it's all about how big the encrypted json representation is, and nothing about the raw memory size (even though they are obviously related)

And off the top of my head, we might as well trim *every* title, even if they would fit? ie, a user with 1 tab with a short URL but a 1.5MB title will fit, but we aren't doing anyone a favor by including all that)